thrust/thrust/system/detail/generic/select_system.h

```diff
@@ -26,90 +26,62 @@
-                                         thrust::detail::minimum_system<System1, System2, System3>>::type&
-select_system(thrust::execution_policy<System1>& system1,
-              thrust::execution_policy<System2>& system2,
-              thrust::execution_policy<System3>& system3);
-
-template <typename System1, typename System2, typename System3, typename System4>
-_CCCL_HOST_DEVICE
-typename thrust::detail::lazy_disable_if<select_system4_exists<System1, System2, System3, System4>::value,
-                                         thrust::detail::minimum_system<System1, System2, System3, System4>>::type&
-select_system(thrust::execution_policy<System1>& system1,
-              thrust::execution_policy<System2>& system2,
-              thrust::execution_policy<System3>& system3,
-              thrust::execution_policy<System4>& system4);
-
-template <typename System1, typename System2, typename System3, typename System4, typename System5>
-_CCCL_HOST_DEVICE typename thrust::detail::lazy_disable_if<
-  select_system5_exists<System1, System2, System3, System4, System5>::value,
-  thrust::detail::minimum_system<System1, System2, System3, System4, System5>>::type&
-select_system(thrust::execution_policy<System1>& system1,
-              thrust::execution_policy<System2>& system2,
-              thrust::execution_policy<System3>& system3,
-              thrust::execution_policy<System4>& system4,
-              thrust::execution_policy<System5>& system5);
-
-template <typename System1, typename System2, typename System3, typename System4, typename System5, typename System6>
-_CCCL_HOST_DEVICE typename thrust::detail::lazy_disable_if<
-  select_system6_exists<System1, System2, System3, System4, System5, System6>::value,
-  thrust::detail::minimum_system<System1, System2, System3, System4, System5, System6>>::type&
-select_system(thrust::execution_policy<System1>& system1,
-              thrust::execution_policy<System2>& system2,
-              thrust::execution_policy<System3>& system3,
-              thrust::execution_policy<System4>& system4,
-              thrust::execution_policy<System5>& system5,
-              thrust::execution_policy<System6>& system6);
+  if constexpr (sizeof...(Systems) == 0)
+  {
+    return thrust::detail::derived_cast(system1);
+  }
+  else if constexpr (sizeof...(Systems) == 1)
+  {
+    // Some glorious hacks to make it work with 2 systems
+    if constexpr ((::cuda::std::is_same_v<System1, Systems> && ...)
+                  || (::cuda::std::is_same_v<System1, thrust::detail::minimum_system_t<System1, Systems>> && ...))
+    {
+      return thrust::detail::derived_cast(system1);
+    }
+    else
+    {
+      static_assert((::cuda::std::is_same_v<Systems, thrust::detail::minimum_system_t<System1, Systems>> && ...));
+      return thrust::detail::derived_cast(systems...);
+    }
+  }
+  else
+  {
+    return select_system(system1, select_system(systems...));
```

Comment:
I am not sure this is correct, because I am not sure whether `select_system(s1, select_system(s2, s3))` is the same as `select_system(select_system(s1, s2), s3)`. The same goes for `minimum_system_t`.